### ISSUE TITLE
Event `onDataChanged` doesn't work even after it is registered

### ISSUE THREAD

**Issue by algomaster99** [OfficeDev/office-js#1231]
## Expected Behavior
When I change the text inside a content control, the `handler` passed inside `contentControls.items[index].onDataChanged(handler)` should **run**. My code snippet for it.

## Current Behavior
I am able to log `Event handler successfully registered` in the console. But whenever I change text inside the current content control;

![cc](https://user-images.githubusercontent.com/35191225/85270332-c287df80-b496-11ea-9ba3-41ab068752c2.PNG)

The handler function doesn't run.

## Steps to Reproduce, or Live Example

The steps for attaching a listener is [here](https://docs.microsoft.com/en-us/office/dev/add-ins/excel/excel-add-ins-events#register-an-event-handler). Although it is for Excel, I assume Word will also work the same way.
```js
const contentControls = context.document.contentControls.getByTag('shipper');
contentControls.load('items/length');
return context.sync().then(() => {
  for(let index=0; index<contentControls.items.length; ++index) {
    contentControls.items[index].onDataChanged.add(handler);
   }
   return context.sync()
      .then(function () {
        console.log("Event handler successfully registered");
      });
});
```
```js
  const handler = (event) => {
    return Word.run((context) => {
        return context.sync()
            .then(() => {
                console.log('event', event)
            });
    });
```

## Context
I am building a Word Add-in which imports a piece of text from some API and marks some words inside in a content control. This is done because I wanted to achieve a functionality where I could change just one of the content control's data and all content controls with the same title changes simultaneously.
![cc](https://user-images.githubusercontent.com/35191225/85271781-dfbdad80-b498-11ea-9eac-7fcc0650efef.PNG)

**Green Boxes:** Content Controls
**Red Box:** The text

Basically, I want to change the text of green boxes whenever one of them is edited. That's why I need `onDataChanged`. Whenever I capture this event, I will get the updated text and then simply [`insertText`](https://docs.microsoft.com/en-us/javascript/api/word/word.contentcontrol?view=word-js-preview#inserttext-text--insertlocation-) into the required content controls.

## Your Environment

* Platform [PC desktop, Mac, iOS, Office Online]: PC Desktop
* Host [Excel, Word, PowerPoint, etc.]: Word
* Office version number: 18.2005.1191.0
* Operating System: Windows
* API version: https://appsforoffice.microsoft.com/lib/beta/hosted/office.js
* Browser (if using Office Online): ______

## Useful logs

- [x] Console errors: None. I get `Event handler successfully registered` which is the desired behaviour
- [ ] Screenshots
- [ ] Test file (if only happens on a particular file)

**Comment by algomaster99** [comment]
I had one more question regarding this issue.
In case I am able to detect the event successfully, and then I make a change to any one of the content controls, the handler will attempt to set the new data in the other content controls with similar `tag`/`title`. When the data is changed for other content controls, _I guess_ it will fire up the `onDataChanged` again (since it is attached to all the content controls with similar `tag`/`title` because a user can change any content control he/she wants) and result in an infinite loop.

By the way, I tried to do this by not using this method. Instead I tried attaching events like this (mixing Office and Word APIs):

```js
Office.context.document.bindings.addFromNamedItemAsync(`Shipper${res+1}`, 'text', { id: `Shipper${res+1}` }, (res) => {
   res.value.addHandlerAsync(Office.EventType.BindingDataChanged, handler);
 });
```

My content controls:
| S. No.  | Title  | Tag  |
|:-:|:-:|:-:|
| 1 | Shipper1  |  shipper |
| 2 | Shipper2  |  shipper |
| 3 | Shipper3  |  shipper |

Here is the handler function:
```js
const handler = (event) => {
    const { binding } = event;
    const bindingId = binding.id; // binding id
    binding.getDataAsync((result) => {
      const data = result.value; // text inside the binding of content control
      Word.run(async context => {
        const contentControl = context.document.contentControls.getByTitle(bindingId).getFirst(); // get the first content control
        contentControl.load('tag'); // load the tag property
        await context.sync();
        const tag = contentControl.tag; // get tag of content control
        const contentControls = context.document.contentControls.getByTag(tag); // get all content controls by Tag
        contentControls.load('items/length');
        await context.sync();
        for(let index=0; index<contentControls.items.length; ++index) {
          contentControls.items[index].insertText(data, Word.InsertLocation.replace); // replace the new text
        }
        return context.sync();
      })
    })
  }
```

**Comment by JuaneloJuanelo** [comment]
Hello @algomaster99, there are  actually 2 questions on this post. so lets take it one by one. 

1. Regarding onDataChanged.add(handler);  -> This is a nice way to subscribe to events, however that's a preview API(you are actually using the preview CDN for office.js and we are not servicing it, so for starters I  recommend you to switch to the production one **https://appsforoffice.microsoft.com/lib/1/hosted/office.js** ) , we  have not released it and the timeline for release its unknown. I am not surprised is not working and probably will behave weirdly in other platforms.  On your second post, you actually found **THE** recommended way to do it: create a content control with a title, then create a binding using the bindings.addFromNamedItemAsync passing the title  and finally once the binding is created you can subscribe to the data changed event. Here is a [working script lab snippet](https://gist.github.com/JuaneloJuanelo/8afa154b336f874849c2c3d29fb75f9f)  that basically demonstrates the flow, it creates a content control on the first paragraph and then the binding and the event subscription.   So that solves the event triggering for sure. 

2.  Regarding the potential infinite loop. Yes if you dont implement this carefully you could end up  in that situation. The mechanism to keep many content control synced with content is to map each content control  the the same XML part. (xml part is literally an XML doc saved within the Word document!) Once an XML part is mapped then bidirectional changes are applied meaning changes in one content control are mapped to the XML part and from there to the other potentially bound ccs. Thats the correct way to do this, but unfortunately we dont have APIs to establish a mapping AND XML mappings are not supported in Word Online.

So, for now what you need to do is that in your handlers you will need to UNSUBSRIBE all the data changed events of the bindings (using binding.removeHandlerAsync) then you need to update the needed content controls and finally  re-establish the events for each affected binding.  Yeah, I know its not the best, but that's the way to do it as of now for your scenario. 

Give it a try! let us know outcomes. this message will autoclose in a few days if we dont hear back from you.

**Comment by algomaster99** [comment]
Hi! So I have made it work what I wanted to do using four steps:
1. Attach content controls to `Bindings`.
2. Whenever a change is triggered in a `content control`, I invoke a `Office.EventType.BindingDataChanged` event listener which gets all the content controls with same `tag`s.
3. I iterate over the all these content controls and update their values. Now the catch is that whenever their values are updated, even programmatically, the event listener is invoked. To prevent infinite loop, I have checked that whenever the values of all the content controls (with same tags) are same, break the loop.

I can't unsubscribe to the changes as there can be multiple edits later on so they have to be there. A small demonstration is as follows:
![demo](https://user-images.githubusercontent.com/35191225/87673828-c1468b80-c792-11ea-9cfc-4667c5999e58.gif)
